### ISSUE TITLE
Chunk offset question

### ISSUE THREAD

**Issue by betteridiot** [samtools/hts-specs#303]
I have been manually creating a pure Python port of all BAM-related random access operations. In doing this, it is requiring me to get into the weeds of the BAM/BAI format. I know my unpacking of the BAI file is correct because I am working with known data. After programmatic unpacking of the refs, bins, and chunks; I am left with exactly the right amount of bytes left to process n_no_coord. 

However, given the first ref, first bin, and first chunk of the file; I get a virtual offset of:
`voffset_beg`: 42049986148
Upon splitting this offset, I get the following:
`coffset_beg`: 641631
`uoffset_beg`: 56932

Additionally, I programmed a BGZF block parser (much like Peter Cock @peterjc does in Bio.bgzf), and iteratively went through the file one block at a time, making a set of all `bgzf_start` of each block...essentially just `tell()` at each block start before unpacking. Each of these blocks was tested for Payload, CM, IDs, and FLG to ensure they met spec. These offsets should be equivalent to all possible `coffset_beg` (from what I assume).

Using the `coffset_beg` shown above, I checked to see if it was in the set of `bgzf_start` offsets and it was not. Furthermore, I scaled this and pulled out ALL `coffset_beg` for ALL chunks, and only a relatively small handful intersected with the `bgzf_start` offsets. If I only got the first one right, I would expect an off by one error. However, the first one wasn't, but some were.

TL;DR
What I am asking is if the `coffset_beg` for each chunk is guaranteed to point at the beginning of its containing BGZF block?

Note: 
OS: 64-bit Windows & Linux
samtools version (for control): Version: 0.1.19-96b5f2294a
Python: 3.6 & 2.7

**Comment by jkbonfield** [comment]
The compressed offset is meant to point to the start of a BGZF block, which appears to be what the code does (although admittedly it's hard to follow).

Are they close to your own recorded start points?  If so, are they always out by the same amount?  It's possible this is simply a difference between the start of the gzip block vs the start of the deflate block.  A BGZF block is a complete gzip file, so it's the former that it refers to.

Note the bins don't have to be in any sensible order, so the first virtual offset may be for a bin which isn't the left-most bin in that reference.

**Comment by peterjc** [comment]
If it is the BAM index (BAI) BIN numbers you're stuck on, have a look at this branch of mine:
https://github.com/peterjc/biopython/blob/SamBam2015/Bio/Sequencing/SamBam/bai.py

I haven't updated this in some years and the BAM format has changed since, but these fundamentals remain the same.

**Comment by betteridiot** [comment]
**UPDATE**: I feel a little stupid now. My BAI parser uses generator functions to parse each of the elements of the BAI format. For some reason, when delegating to the generator in a `yield from` statement, the generators were no longer guaranteed to be synchronized within the outer scope...or maybe some other weird gotcha about generators. Anyways, because of that, the generators were not following the head in the right order. Once I changed it to straight `return` statements instead, all my coffsets aligned with the BGZF block offset. Sorry about the trouble, and feel free to close.

**Comment by peterjc** [comment]
I'm glad you could solve it, and that my implementation was useful.

**Comment by jkbonfield** [comment]
No problem and glad you have it fixed.
